# Map, Filter, Reduce, Lambda & Recursion

## Tasks Today:

1) <b>Lambda Functions</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Saving to a Variable <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Multiple Inputs <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Passing a Lambda into a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; e) Returning a Lambda from a Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; f) In-Class Exercise #1 <br>
2) <b>Map</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Map <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #2 <br>
3) <b>Filter</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Filter <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #3 <br>
4) <b>Reduce</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Syntax <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Using Lambda's with Reduce <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #4 <br>
5) <b>Recursion</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Implementing a Base <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Writing a Factorial Function <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #5 <br>
6) <b>Generators & Iterators</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Yield Keyword <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Inifinite Generator <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) In-Class Exercise #6 <br>
7) <b>Exercises</b> <br>
 &nbsp;&nbsp;&nbsp;&nbsp; a) Exercise #1 - Filtering Empty Strings <br>
 &nbsp;&nbsp;&nbsp;&nbsp; b) Exercise #2 - Sorting with Last Name <br>
 &nbsp;&nbsp;&nbsp;&nbsp; c) Exercise #3 - Conversion to Farhenheit <br>
 &nbsp;&nbsp;&nbsp;&nbsp; d) Exercise #4 - Fibonacci Sequence <br>

## Lambda Functions <br>
<p>Lambda functions... or "Anonymous Functions" are referring to inline functions with no name. The keyword lambda denotes the no name function, and executes within a single line. Without saving it to a variable; however, it is not able to be used, unless passed in either as a paramater or within list comprehension.<br>Written as "(keyword lambda) (one or more inputs) (colon) (function to be executed)"</p>

#### Syntax

In [1]:
# User Defined Function
def add_two(num):
    return num + 2

add_two(4)

6

In [6]:
# Same function as a lambda function
(lambda num: num + 2)

<function __main__.<lambda>(num)>

#### Saving to a Variable

In [8]:
f_test = lambda num: num + 2


f_test(5)

7

#### Multiple Inputs

In [10]:
# Multiple inputs with no variable
(lambda x, y, z: x * y * z)(2, 5, 7)

70

In [15]:
# Multiple inputs with a variable attached
mult_func = lambda x, y, z: x * y * z

mult_func(3, 4, 6)


72

#### Passing a Lambda into a Function

In [22]:
def apply_func(f, num):
    """
    f expects a function
    num expects a number
    apply_func will run function f and pass in num as an argument
    """
    
    return f(num)

# def my_func(x):
#     return x * x

apply_func(lambda x: x*x, 4) # lambda 4: 4 * 4 = 16

16

#### Returning a Lambda from a Function

In [36]:
# Regular defined function returns int
def multiply_test(num):
    return num * 4

type(multiply_test(3))

# Function that returns a function
def return_func():
    def multiply(num):
        return num * 2
    return multiply


a = return_func()

a(5)

print(return_func)
print(return_func())


# Lambda function returned from a regular function
def return_lambda(b, c):
    return lambda x: x * (b + c)

r_lamb = return_lambda(3, 4)
print(r_lamb)
r_lamb(5)

<function return_func at 0x00000275BF3C2DC0>
<function return_func.<locals>.multiply at 0x00000275BF3EF0D0>
<function return_lambda.<locals>.<lambda> at 0x00000275BF3EF3A0>


35

#### If Statements within Lambdas

In [39]:
# lambda x: Return value True if (condition) else false return value
f_condition = lambda num: num * 2 if num > 10 else num + 2

print(f_condition(8))

print(f_condition(12))

print(f_condition(10))

10
24
12


In [42]:
new_test = lambda num: num * 2 if num > 20 else num * 3 if num < 5 else num ** 2

print(new_test(22))
print(new_test(2))
print(new_test(10))

44
6
100


#### In-Class Exercise #1 <br>
<p>Write an anonymous function that cubes the arguments passed in and assign the anonymous function to a variable 'f'.</p>

In [48]:
f = lambda brianstanton: brianstanton ** 3

print(f(3))
print(f(5))

27
125


## Map <br>
<p>The map function allows you to iterate over an entire list while running a function on each item of the list. This is why the map function works well with lambda's, because it simplifies things and you write less lines of code.<br>The syntax for a map function is "map(function to be used, list to be used)"<br>However, you must be careful, as the map function returns a map object, not a list. To turn it into a list we use the list() type conversion.</p>

#### Syntax

In [73]:
def square(num):
    return num ** 2

list(map(square, [1, 2, 3, 4]))

[1, 4, 9, 16]

In [70]:
# map(function_to_apply, iterable_function_applied_to(list, tuple, set, dict..))

# With a user defined function
def square(num, num2):
    if num < 10 and num2 < 10:
        return num **2, num2 ** 2
    else:
        return num, num2

numbers = [4, 22, 44, 2, 4, 8, 3, 66]
more_numbers = [2, 10, 22, 7]

squared_nums = list(map(square, numbers, more_numbers))
squared_nums

[(16, 4), (22, 10), (44, 22), (4, 49)]

#### Using Lambda's with Map

In [71]:
# map(lambda func, iterable)

squared_lamb_nums = map(lambda num, num2: (num**2, num2**2)if num < 10 and num2 < 10 else (num, num2), numbers, more_numbers)
list(squared_lamb_nums)

[(16, 4), (22, 10), (44, 22), (4, 49)]

#### In-Class Exercise #2 <br>
<p>Use the map function to double each number and minus it by one in the list by using a lambda function</p>

In [72]:
my_nums = [5, 8, 10, 15, 24, 42, 17, 14, 3, 9]

list(map(lambda num: num*2 - 1,my_nums))

[9, 15, 19, 29, 47, 83, 33, 27, 5, 17]

## Filter() <br>
<p>Filter's are similar to the map function, where you're able to pass a function argument and a list argument and filter out something from the list based on the conditions passed. Similar to the map function, it returns a filter object, so you need to type convert it to a list()</p>

#### Syntax

In [78]:
# filter(function_to_apply, iterable_function_applied_to(list, tuple, set, dict..))

def is_even(num):
    if num % 2 == 0:
        return True
    else:
        return False
    
list_of_nums = [1, 5, 2, 6, 8, 33, 6, 2, 55, 24, 64, 234, 432, 42]

list(filter(is_even, list_of_nums))

[2, 6, 8, 6, 2, 24, 64, 234, 432, 42]

In [82]:
def names_that_start_with_a(name):
    if name[0].lower() == 'a':
        return True
    return False

names = ['Bobby', 'Andrew', 'Angelica', 'Frank', 'Amber', 'Philip']

list(filter(names_that_start_with_a, names))

['Andrew', 'Angelica', 'Amber']

In [90]:
def truthyness(num):
    if num > 10:
        return 'Hello World'
    else:
        return ''
    
    
list(filter(truthyness, [1, 2, 3, 15, 16]))

[15, 16]

#### Using Lambda's with Filter()

In [92]:
my_name_here = 'Arian'
my_name_here[0].lower() == 'a'

True

In [97]:
new_names_lamb = list(filter(lambda name: name[0].lower() == 'a', names))
new_names_lamb

['Andrew', 'Angelica', 'Amber']

#### In-Class Exercise #3 <br>
<p>Filter out all the numbers that are below the mean of the list.<br><b>Hint: Import the 'statistics' module</b></p>

In [101]:
import statistics

statistics.mean([2, 2, 2, 4, 4, 4])

3

In [116]:
my_nums = [5, 8, 10, 15, 24, 42, 17, 14, 3, 9]

mean = statistics.mean(my_nums)

list(filter(lambda num: num > mean, my_nums))

[15, 24, 42, 17]

In [115]:
my_nums = [5, 8, 10, 15, 24, 42, 17, 14, 3, 9]

mean = statistics.mean(my_nums)

def is_greater_than_mean(num):
    if num > mean:
        return True
    else:
        return False

list(filter(is_greater_than_mean, my_nums))

[15, 24, 42, 17]

## Reduce() <br>
<p>Be very careful when using this function, as of Python 3 it's been moved to the 'functools' library and no longer is a built-in function.<br>The creator of Python himself, says to just use a for loop instead.</p>

#### Syntax

In [117]:
from functools import reduce

# reduce(function, iterable)

def add_nums(accumulated_value, update_value):
    return accumulated_value + update_value


list_1 = [2, 4, 6, 8, 10]

total = 0
for num in list_1:
    total = add_nums(total, num)
print(total)


30


In [122]:
help(reduce)

Help on built-in function reduce in module _functools:

reduce(...)
    reduce(function, sequence[, initial]) -> value
    
    Apply a function of two arguments cumulatively to the items of a sequence,
    from left to right, so as to reduce the sequence to a single value.
    For example, reduce(lambda x, y: x+y, [1, 2, 3, 4, 5]) calculates
    ((((1+2)+3)+4)+5).  If initial is present, it is placed before the items
    of the sequence in the calculation, and serves as a default when the
    sequence is empty.



In [118]:
reduce(add_nums, list_1)

30

In [119]:
def subtract_nums(acc, upd):
    return acc - upd


reduce(subtract_nums, list_1)

-26

#### Using Lambda's with Reduce()

In [120]:
my_list = [1, 2, 3, 4, 5]

reduce(lambda x,y: x * y , my_list)

120

In [121]:
total = 1
for num in my_list:
    total *= num
    
total

120

#### In-Class Exercise #4 <br>
<p>Use the reduce function to multiply the numbers in the list below together with a lambda function.</p>

In [123]:
my_list = [1, 2, 3, 4, 5]

reduce(lambda x,y: x * y, my_list)

120

## Recursion <br>
<p>Recursion means that a function is calling itself, so it contanstly executes until a base case is reached. It will then push the returning values back up the chain until the function is complete. A prime example of recursion is computing factorials... such that 5! (factorial) is 5*4*3*2*1 which equals 120.</p>

#### Implementing a Base Case

In [128]:
def add_nums(num):
    # Set base case for recursive function
    if num <= 1:
        print("add_nums(1) = 1")
        return num
    else:
        print(f"add_nums({num}) = {num} + add_nums({num - 1})")
        return num + add_nums(num - 1)
    
add_nums(5)

add_nums(5) = 5 + add_nums(4)
add_nums(4) = 4 + add_nums(3)
add_nums(3) = 3 + add_nums(2)
add_nums(2) = 2 + add_nums(1)
add_nums(1) = 1


15

In [125]:
def add_two(num):
    print(f"add_two({num}) = {num + 2}")
    return num + 2

add_two(5)

add_two(5) = 7


7

#### Writing a Factorial Function

In [136]:
# 5! = 5 * 4!
# 4! = 4 * 3!
# 3! = 3 * 2!
# 2! = 2 * 1!
# 1! = 1

def factorial(num):
    if num <= 1:
        return num
    else:
        return num * factorial(num - 1)
    
factorial(5)

120

#### In-Class Exercise #5 <br>
<p>Write a recursive function that subtracts all numbers to the argument given.</p>

In [138]:
def sub_nums(num):
    if num <= 1:
        print(f'sub_nums(1) = 1')
        return num
    else:
        print(f'sub_nums({num}) = {num} - sub_nums({num-1})')
        return num - sub_nums(num - 1)
    
sub_nums(10)

sub_nums(10) = 10 - sub_nums(9)
sub_nums(9) = 9 - sub_nums(8)
sub_nums(8) = 8 - sub_nums(7)
sub_nums(7) = 7 - sub_nums(6)
sub_nums(6) = 6 - sub_nums(5)
sub_nums(5) = 5 - sub_nums(4)
sub_nums(4) = 4 - sub_nums(3)
sub_nums(3) = 3 - sub_nums(2)
sub_nums(2) = 2 - sub_nums(1)
sub_nums(1) = 1


5

In [139]:
class Example:
    def __init__(self, value):
        self.value = value
        self.next = None
        
a = Example(1)
b = Example(2)
c = Example(3)

a.next = b
b.next = c


def print_last_item_of_linked_list(head_of_linked_list):
    if not head_of_linked_list.next:
        print(head_of_linked_list.value)
    else:
        print_last_item_of_linked_list(head_of_linked_list.next)
        
print_last_item_of_linked_list(a)

3


## Generators <br>
<p>Generators are a type of iterable, like lists or tuples. They do not allow indexing, but they can still be iterated through with for loops. They are created using functions and the yield statement.</p>

#### Yield Keyword <br>
<p>The yield keyword denotes a generator, it doesn't return so it won't leave the function and reset all variables in the function scope, instead it yields the number back to the caller.</p>

In [167]:
def my_range(stop, start=0, step=1):
    while start <= stop:
        yield start
        start += step
        
test = my_range(10)

In [168]:
for i in test:
    print(i)

0
1
2
3
4
5
6
7
8
9
10


In [150]:
list(test)

[]

#### Infinite Generator

In [ ]:
# bad, never create infinite loops

#### In-Class Exercise #6 <br>
<p>Create a generator that takes a number argument and yields that number squared, then prints each number squared until zero is reached.</p>

In [170]:
def nums_squared(num):
    while num >= 0:
        yield num ** 2
        num -= 1
        
        
for s in nums_squared(10):
    print(s)

100
81
64
49
36
25
16
9
4
1
0


# Exercises

### Exercise #1 <br>
<p>Filter out all of the empty strings from the list below</p>

`Output: ['Argentina', 'San Diego', 'Boston', 'New York']`

In [ ]:
places = [" ","Argentina", " ", "San Diego","","  ","","Boston","New York"]



### Exercise #2 <br>
<p>Write an anonymous function that sorts this list by the last name...<br><b>Hint: Use the ".sort()" method and access the key"</b></p>

`Output: ['Victor aNisimov', 'Gary A.J. Bernstein', 'Joel Carter', 'Andrew P. Garfield', 'David hassELHOFF']`

In [ ]:
author = ["Joel Carter", "Victor aNisimov", "Andrew P. Garfield","David hassELHOFF","Gary A.J. Bernstein"]



### Exercise #3 <br>
<p>Convert the list below from Celsius to Farhenheit, using the map function with a lambda...</p>

`Output: [('Nashua', 89.6), ('Boston', 53.6), ('Los Angeles', 111.2), ('Miami', 84.2)]
`

In [ ]:
# F = (9/5)*C + 32
places = [('Nashua',32),("Boston",12),("Los Angeles",44),("Miami",29)]


### Exercise #4 <br>
<p>Write a recursion function to perform the fibonacci sequence up to the number passed in.</p>

`Output for fib(5) => 
Iteration 0: 1
Iteration 1: 1
Iteration 2: 2
Iteration 3: 3
Iteration 4: 5
Iteration 5: 8`